hogben/models/base.py

```diff
@@ -56,16 +55,61 @@ class BaseSample(VariableAngle):
     """Abstract class representing a "standard" neutron reflectometry sample
     defined by a series of contiguous layers."""
 
-    @abstractmethod
-    def sld_profile(self):
-        """Plots the SLD profile of the sample."""
-        pass
+    def __init__(self):
+        """Initialise the sample and define class attributes"""
+        self._structures = []
 
     @abstractmethod
-    def reflectivity_profile(self):
-        """Plots the reflectivity profile of the sample."""
+    def get_structures(self):
+        """
+        Get a list of the possible sample structures.
+        """
         pass
 
+    @property
+    def structures(self):
+        """Return the structures that belong to the sample"""
+        return self.get_structures()
+
+    @structures.setter
+    def structures(self, structures):
+        self._structures = structures
+
+    def get_param_by_attribute(self, attr: str) -> list:
+        """
+        Get all parameters defined in the sample model that have a given
+        attribute. Returns a list with all parameters with this attribute,
+        e.g. `attr='vary'` returns all varying parameters.
+
+        Args:
+            attr (str): The attribute to filter for
+
+        Returns:
+            list: A list of all parameters with the given attribute
+        """
+        params = []
+        for model in self.get_models():
+            for p in flatten(model.parameters):
+                if hasattr(p, attr) and getattr(p, attr):
+                    params.append(p)
+                    continue
+                if p._deps:
+                    params.extend([_p for _p in p.dependencies() if
+                                   hasattr(_p, attr) and getattr(_p, attr)])
+        return list(set(params))
+
+    def get_models(self) -> list:
+        """
+        Returns a list of all refnx `ReflectModel` models that are
```

Comment:
I've changed the docstring. 

For the lipids, I've added a seperate implementation for `get_models`  in `BaseLipids`, where it takes advantage of the fact that these structures all have a separately defined scale and bkg. 

What I could do as well, is to modify the basic Sample (`samples.Sample`), such that it also has a list of `bkg`, `dq' and `scale`, corresponding to the total amount of structures. If a single value is given, it would then just create a list with the same values. And if a list is given it would use the list. If the total amount of elements do not correspond to the total amount of structures, an error should be raised. Will look into that before  resolving. That could deduplicate things a bit further.

Reply:
So I've changed it a bit more. Now as described, `dq`, `scale` and `bkg` can either be given as a single float/int. In that case the same value will be used for each structure. Or it can be given as a list, in case it will use the specified values in the same order as the structures.

For the magnetic cases (PR after this), I will likely keep it so that `scale`, `dq` and `bkg` will be the same for spin-up and spin-down. Meaning the same amount of parameters need to be given for polarised experiments as for unpolarised.